libs/partners/ibm/langchain_ibm/llms.py

```diff
@@ -0,0 +1,396 @@
+import logging
+import os
+from typing import Any, Dict, Iterator, List, Mapping, Optional, Union
+
+from langchain_core.callbacks import CallbackManagerForLLMRun
+from langchain_core.language_models.llms import BaseLLM
+from langchain_core.outputs import Generation, GenerationChunk, LLMResult
+from langchain_core.pydantic_v1 import Extra, SecretStr, root_validator
+from langchain_core.utils import convert_to_secret_str, get_from_dict_or_env
+
+logger = logging.getLogger(__name__)
+
+
+class WatsonxLLM(BaseLLM):
+    """
+    IBM watsonx.ai large language models.
+
+    To use, you should have ``ibm_watsonx_ai`` python package installed,
+    and the environment variable ``WATSONX_APIKEY`` set with your API key, or pass
+    it as a named parameter to the constructor.
+
+
+    Example:
+        .. code-block:: python
+
+            from ibm_watsonx_ai.metanames import GenTextParamsMetaNames
+            parameters = {
+                GenTextParamsMetaNames.DECODING_METHOD: "sample",
+                GenTextParamsMetaNames.MAX_NEW_TOKENS: 100,
+                GenTextParamsMetaNames.MIN_NEW_TOKENS: 1,
+                GenTextParamsMetaNames.TEMPERATURE: 0.5,
+                GenTextParamsMetaNames.TOP_K: 50,
+                GenTextParamsMetaNames.TOP_P: 1,
+            }
+
+            from langchain_ibm import WatsonxLLM
+            watsonx_llm = WatsonxLLM(
+                model_id="google/flan-ul2",
+                url="https://us-south.ml.cloud.ibm.com",
+                apikey="*****",
+                project_id="*****",
+                params=parameters,
+            )
+    """
+
+    model_id: str = ""
+    """Type of model to use."""
+
+    deployment_id: str = ""
+    """Type of deployed model to use."""
+
+    project_id: str = ""
+    """ID of the Watson Studio project."""
+
+    space_id: str = ""
+    """ID of the Watson Studio space."""
+
+    url: Optional[SecretStr] = None
```

Comment:
I think some of these can be `str` not `SecretStr`

Reply:
All fields used in credentials we would like keep as `SecretStr`